NEWS.md

```diff
@@ -153,6 +154,19 @@
     #
     # [[3]]
     # [1] 5 1 2 3 4
+
+    # Benchmark
+    x = seq.int(1e9) # 3.8 GB
```

Comment:
This is altrep, so timing suffers from materializing this object. This happens for both base R and data.table so I think we can ignore it. It just may be slower than if you would use altrep+1-1 as the input object.

Reply:
Good point. But altrep expands on the first time though : 
```
> x = seq.int(1e9)
> .Internal(inspect(x))
@55cf162368e8 13 INTSXP g0c0 [REF(65535)]  1 : 1000000000 (compact)
> y = shift(x, 1, type="cyclic")
> .Internal(inspect(x))
@55cf162368e8 13 INTSXP g0c0 [MARK,REF(65535)]  1 : 1000000000 (expanded)
```
So after the first run, the expanded altrep will be used, iiuc. This is why, iiuc, the max of shift() is 3.05s, an outlier and the first run. I confirmed it is the first run by using system.time() 3 times. If the base R way was tested before shift, then its max would also be an outlier, iiuc, although harder to see because the op takes longer so the extra fixed time for the altrep expansion on the first run isn't as big as a percentage.

> This is an altrep, so timing suffers from materializing this object.

Only the first run suffers, which becomes the max.

> This happens for both base R and data.table so I think we can ignore it.

It doesn't happen for both. It happens for the first run of whichever is tested first. But it can be ignored if we ignore the max column.

@ben-schwen To eliminate the risk of suspicion in the mind of the reader of the news item, how about just using `x=sample(1e9)`. That removes the suspicion too that shift() could be using the property that the input is a sequence to not actually read all of the input.

On my laptop:
```
> x = sample(1e9) # 3.8 GB
microbenchmark::microbenchmark(
  shift(x, 1, type="cyclic"),
  c(tail(x, 1), head(x,-1)),
  times = 5L,
  unit = "s"
)
Unit: seconds
                         expr      min       lq     mean   median       uq      max neval
 shift(x, 1, type = "cyclic") 1.427934 1.462322 1.487594 1.469971 1.536985 1.540759     5
   c(tail(x, 1), head(x, -1)) 6.359621 6.575269 6.622709 6.684869 6.746652 6.747135     5
```

Reply:
I'm glad I peppered my statements with "iiuc", always good to do that. I went and double-checked, and actually the base R way does not expand the altrep.

```
> x = seq.int(1e9)
> .Internal(inspect(x))
@55cf16d3bdf0 13 INTSXP g0c0 [REF(65535)]  1 : 1000000000 (compact)
> system.time(y<-c(tail(x, 1), head(x,-1)))
   user  system elapsed 
  6.261   1.948   8.210 
> .Internal(inspect(x))
@55cf16d3bdf0 13 INTSXP g1c0 [MARK,REF(65535)]  1 : 1000000000 (compact)
```

Repeating the system.time gives a stable timing, including the expansion of the altrep each time.  I had been thinking my laptop had faster memory or something than Ben's (which was a bit puzzling why I was seeing 6.7s and he was seeing 9.1s, a big difference).  It actually seems that the expansion of the altrep takes 1.2s and that was unfair to the base R way as that's happening for each and every one of its runs, but was only happening for the first run of shift().

Reply:
I replaced it now with the suggested `x=sample(1e9)` and I agree that shifting with groups is probably the most common use-case. The speedup of 4x `shift` in comparison with a naive implementation is recognizable for vectors of length 1e5 and longer. Using `shift` with many groups is also a known bottleneck #1534 and it appears that `shift` is even worse than a naive implementation (I would guess due to the massive amount of `memmove`s). I'm already working on a version of `gshift`.

## Benchmarks for using shift with many groups (cylic and lag)
```R
N = 1e7
DT = data.table(x = sample(N), y = sample(1e5,N,TRUE))
microbenchmark::microbenchmark(
  DT[, shift(x, 1, type="cyclic"), y],
  DT[, c(tail(x, 1), head(x,-1)), y],
  times = 10L,
  unit = "s"
)
# Unit: seconds
#                                   expr      min       lq     mean   median       uq      max neval
#  DT[, shift(x, 1, type = "cyclic"), y] 1.951673 1.981829 2.010449 2.001052 2.019000 2.118122    10
#    DT[, c(tail(x, 1), head(x, -1)), y] 1.666980 1.675234 1.693656 1.688608 1.712064 1.721558    10

N = 1e7
DT = data.table(x = sample(N), y = sample(1e5,N,TRUE))
microbenchmark::microbenchmark(
  DT[, shift(x, 1, type="lag"), y],
  DT[, c(NA, head(x,-1)), y],
  times = 10L,
  unit = "s"
)
Unit: seconds
                               expr      min       lq     mean   median       uq      max neval
 DT[, shift(x, 1, type = "lag"), y] 1.888065 1.926722 2.036473 1.991713 2.180608 2.292505    10
        DT[, c(NA, head(x, -1)), y] 1.138030 1.156684 1.203427 1.179590 1.234704 1.338751    10

```